dev/fattest.simplicity/src/componenttest/security/utils/SSLUtils.java

```diff
@@ -0,0 +1,175 @@
+import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.bouncycastle.operator.ContentSigner;
+import org.bouncycastle.operator.OperatorCreationException;
+import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
+import org.bouncycastle.openssl.jcajce.JcaPEMWriter;
+
+import com.ibm.websphere.simplicity.log.Log;
+
+public class SSLUtils {
+
+    private static final Class<?> c = SSLUtils.class;
+
+    public static KeyPair generateKeyPair() throws Exception {
+        KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA");
+        generator.initialize(2048, new SecureRandom());
+        KeyPair pair = generator.generateKeyPair();
+
+        return pair;
+    }
+
+    public static Certificate selfSign(KeyPair keyPair, String subjectDN, List<String> genericNameList)
+            throws Exception
+    {
+        Log.info(c, "selfSign", "entering SSLUtils.selfSign");
+
+        final String DNS_PATTERN = "^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\\-]*[A-Za-z0-9])$";
+        final String IPV4_PATTERN = "^(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])(\\.(?!$)|$)){4}$";
+
+        Provider bcProvider = new BouncyCastleProvider();
+        Security.addProvider(bcProvider);
+
+        long now = System.currentTimeMillis();
+        Date startDate = new Date(now);
+
+        X500Name dnName = new X500Name(subjectDN);
+
+        // Using the current timestamp as the certificate serial number
+        BigInteger certSerialNumber = new BigInteger(Long.toString(now));
+
+        Calendar calendar = Calendar.getInstance();
+        calendar.setTime(startDate);
+        // 1 Yr validity
+        calendar.add(Calendar.YEAR, 1);
+
+        Date endDate = calendar.getTime();
+
+        // Use appropriate signature algorithm based on your keyPair algorithm.
+        String signatureAlgorithm = "SHA256WithRSA";
+
+        SubjectPublicKeyInfo subjectPublicKeyInfo = SubjectPublicKeyInfo.getInstance(keyPair
+                .getPublic().getEncoded());
+
+        X509v3CertificateBuilder certificateBuilder = new X509v3CertificateBuilder(dnName,
+                certSerialNumber, startDate, endDate, dnName, subjectPublicKeyInfo);
+
+        List<GeneralName> altNames = new ArrayList<GeneralName>();
+        for (String name : genericNameList) {
+            if (name.matches(DNS_PATTERN)) {
+                Log.info(c, "selfSign", "adding dns GeneralName " + name);
+                altNames.add(new GeneralName(GeneralName.dNSName, name));
+            } else if (name.matches(IPV4_PATTERN)) {
+                Log.info(c, "selfSign", "adding ip GeneralName " + name);
+                altNames.add(new GeneralName(GeneralName.dNSName, name));
+            }
+        }
+
+        GeneralNames subjectAltNames = GeneralNames.getInstance(new DERSequence((GeneralName[]) altNames.toArray(new GeneralName[] {})));
```

Comment:
This line looks weird, does it need the cast and the `DERSequence` in there?

Reply:
I would have to research to find out, the javadoc is terrible for bouncycastle. getInstance takes java.lang.Object and the javadoc says nothing about what it really wants.